### PR TITLE
[Cosmos] Adds full text policy and full text indexes

### PR DESCRIPTION
**Pull request by simorenoh**
This PR adds the functionality, samples and tests for creating and replacing containers with full text policies and full text indexes for the SDK.

A sample full text policy looks like this:
```python
full_text_policy = {
    "defaultLanguage": "en-US",
    "fullTextPaths": [
        {
            "path": "/text",
            "language": "en-US"
        }
    ]
}
```

For the indexing bit, we now simply accept `fullTextIndexes` to the existing indexing policy:
```python
indexing_policy = {
    "automatic": True,
    "indexingMode": "consistent",
    "compositeIndexes": [
        [
            {"path": "/numberField", "order": "ascending"},
            {"path": "/stringField", "order": "descending"}
        ]
    ],
    "fullTextIndexes": [
        {"path": "/text"}
    ]
}
```

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/README.md [hunk 1]
```diff
@@ -750,6 +750,52 @@ not being able to recognize the new NonStreamingOrderBy capability that makes ve
 If this happens, you can set the `AZURE_COSMOS_DISABLE_NON_STREAMING_ORDER_BY` environment variable to `"True"` to opt out of this
 functionality and continue operating as usual.*
 
+### Public Preview - Full Text Policy and Full Text Indexes
+We have added new capabilities to utilize full text policies and full text indexing for users to leverage full text search
+utilizing our Cosmos SDK. These two container-level configurations have to be turned on at the account-level
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/README.md:755]: Link to how to enable them at account-level is missing?

File: sdk/cosmos/azure-cosmos/README.md [hunk 2]
```diff
@@ -750,6 +750,52 @@ not being able to recognize the new NonStreamingOrderBy capability that makes ve
...
+before you can use them.
+
+A full text policy allows the user to define the default language to be used for all full text paths, or to set
+a language for each path individually in case the user would like to use full text search on data containing different
+languages in different fields.
+
+A sample full text policy would look like this:
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/README.md:762]: what format is the lnaguage expected to be in I assume <L>-<C> (with <L> being the ISO-639 lanugauge code and <C> the ISO-3166 Country code? Are there plans to support some special-casing for neutral or similar? What about use cases where  a customer does not know the exact language - can they just use en (wnglish independent of country)?

File: sdk/cosmos/azure-cosmos/README.md [hunk 3]
```diff
@@ -750,6 +750,52 @@ not being able to recognize the new NonStreamingOrderBy capability that makes ve
...
+            "language": "en-US"
+        },
+        {
+            "path": "/text2",
+            "language": "en-US"
+        }
+    ]
+}
+```
+Currently, the only supported language is `en-US`.
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/README.md:778]: Supported is one question - what happens when another language is chosen? Error? Language ignored and handled like en-US? What are the plans to extend language support? Never? Long-term? short-term?

File: sdk/cosmos/azure-cosmos/samples/index_management.py [hunk 4]
```diff
@@ -746,6 +746,61 @@ def get_embeddings(num):
         print("Entity doesn't exist")
 
 
+def use_full_text_policy(db):
+    try:
+        delete_container_if_exists(db, CONTAINER_ID)
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/samples/index_management.py:751]: Deleting a container is happening async - espcially on multi-region account you probably need to add a dely to not introduce flakiness before trying to add a new container (or handle the 409 there and retry until new ocntainer can be created)

**Review by FabianMeiswinkel — CHANGES_REQUESTED** [review]
LGTM - but left few comments that need to be addressed before merging.

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_database.py [hunk 5]
```diff
@@ -509,6 +521,9 @@ async def replace_container(
             note that analytical storage can only be enabled on Synapse Link enabled accounts.
         :keyword response_hook: A callable invoked with the response metadata.
         :paramtype response_hook: Callable[[Dict[str, str], Dict[str, Any]], None]
+        :keyword Dict[str, Any] full_text_policy: **provisional** The full text policy for the container.
```
[the remark this answers is not in the document]
Comment (Pilchie) [reply sdk/cosmos/azure-cosmos/azure/cosmos/aio/_database.py:524]: The policy should be atomic, but you are right that adding an index won't be.

File: sdk/cosmos/azure-cosmos/README.md (hunk 1 above)
Comment (simorenoh) [reply sdk/cosmos/azure-cosmos/README.md:755]: based on the way we did the vector search release, we did not share the link there - not sure if it was available and not sure if this one is either

File: sdk/cosmos/azure-cosmos/README.md (hunk 3 above)
Comment (simorenoh) [reply sdk/cosmos/azure-cosmos/README.md:778]: this is a public document that does not exist yet afaik - I'll go ahead and update on the specifics I can do at the moment

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_database.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/cosmos/azure-cosmos/azure/cosmos/aio/_database.py:524]: I'll make a note on the index update taking time in the README documentation then

File: sdk/cosmos/azure-cosmos/samples/index_management.py (hunk 4 above)
Comment (simorenoh) [reply sdk/cosmos/azure-cosmos/samples/index_management.py:751]: this is just a snippet we use for samples, and every other method in this file does the same thing - I can go ahead and just change the id if you think there's an issue, but it's the general pattern we have across the file at the moment

File: sdk/cosmos/azure-cosmos/test/test_full_text_policy.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+            indexing_policy=indexing_policy
+        )
+        properties = created_container.read()
+        assert properties["fullTextPolicy"] == full_text_policy
+        assert properties["indexingPolicy"]['fullTextIndexes'] == indexing_policy['fullTextIndexes']
+
+        # Replace the container with new policies
+        full_text_policy['fullTextPaths'][0]['path'] = "/new_path"
+        indexing_policy['fullTextIndexes'][0]['path'] = "/new_path"
+        replaced_container = self.test_db.create_container(
```
Comment (allenkim0129) [inline sdk/cosmos/azure-cosmos/test/test_full_text_policy.py:102]: In the doc of `create_container`, it says "If a container with the given ID already exists, a CosmosResourceExistsError is raised." 

Is it not true anymore? 

I don't see any change enabling `replace` behavior

File: sdk/cosmos/azure-cosmos/test/test_full_text_policy.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+            id=created_container.id,
+            partition_key=PartitionKey(path="/id"),
+            full_text_policy=full_text_policy,
+            indexing_policy=indexing_policy
+        )
+        properties = replaced_container.read()
```
Comment (allenkim0129) [inline sdk/cosmos/azure-cosmos/test/test_full_text_policy.py:108]: If container has been replaced, if we call the same test with original `created_container`, would we get the same result?

File: sdk/cosmos/azure-cosmos/test/test_full_text_policy.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+        try:
+            self.test_db.create_container(
+                id='full_text_container',
+                partition_key=PartitionKey(path="/id"),
+                full_text_policy=full_text_policy_no_langs
+            )
+            pytest.fail("Container creation should have failed for lack of embedding policy.")
+        except exceptions.CosmosHttpResponseError as e:
+            assert e.status_code == 400
+            assert "The Full Text Policy contains invalid syntax" in e.http_error_message
```
Comment (allenkim0129) [inline sdk/cosmos/azure-cosmos/test/test_full_text_policy.py:153]: Where these error messages constructed? Can we update the error message with more details? Users many do not know exactly what was missing.

File: sdk/cosmos/azure-cosmos/test/test_full_text_policy.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+        try:
+            self.test_db.create_container(
+                id='full_text_container',
+                partition_key=PartitionKey(path="/id"),
+                indexing_policy=indexing_policy_wrong_path
+            )
+            pytest.fail("Container creation should have failed for lack of embedding policy.")
+        except exceptions.CosmosHttpResponseError as e:
+            assert e.status_code == 400
+            assert "Full-text index specification at index (0) contains invalid path" in e.http_error_message
```
Comment (allenkim0129) [inline sdk/cosmos/azure-cosmos/test/test_full_text_policy.py:243]: Nit: Why we have completely different error message format with similar issue from `full text policy` ? Can we make them more uniformed? 

"The Full Text Policy contains an invalid Path: abstract"

Comment (simorenoh) [reply sdk/cosmos/azure-cosmos/test/test_full_text_policy.py:263]: all of these errors come from the service in the backend - we don't fabricate any of them in the SDK. these checks are here to validate that they stay the same, or that if they change we will become aware of it

File: sdk/cosmos/azure-cosmos/test/test_full_text_policy.py (hunk 8 above)
Comment (simorenoh) [reply sdk/cosmos/azure-cosmos/test/test_full_text_policy.py:173]: it's all in the service (backend) - we have already provided this feedback to them, but yep completely agreed. this is just what they are for now

File: sdk/cosmos/azure-cosmos/test/test_full_text_policy.py (hunk 7 above)
Comment (simorenoh) [reply sdk/cosmos/azure-cosmos/test/test_full_text_policy.py:128]: nope - that's why the previous call to `container.read()` checks that the properties match those of the first container and not the second one. I'll add a check comparing the two against each other too.

File: sdk/cosmos/azure-cosmos/test/test_full_text_policy.py (hunk 6 above)
Comment (simorenoh) [reply sdk/cosmos/azure-cosmos/test/test_full_text_policy.py]: yep that was my bad, used the wrong method

File: sdk/cosmos/azure-cosmos/test/test_full_text_policy_async.py [hunk 10]
```diff
@@ -0,0 +1,274 @@
+# The MIT License (MIT)
+# Copyright (c) Microsoft Corporation. All rights reserved.
+
+import unittest
+import uuid
+
```
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/cosmos/azure-cosmos/test/test_full_text_policy_async.py:6]: done

**Review by allenkim0129 — APPROVED** [review]

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - Thanks!

**Pull request merged**